fsspec/dircache.py

```diff
@@ -0,0 +1,50 @@
+from functools import lru_cache
+import time
+
+
+class DirCache(dict):
+    """
+    Caching of directory listings, in a structure like
+
+    {"path0": [
+        {"name": "path0/file0",
+         "size": 123,
+         "type": "file",
+         ...
+        },
+        {"name": "path0/file1",
+        },
+        ...
+        ],
+     "path1": [...]
+    }
+
+    Parameters to this class control listing expiry or indeed turn
+    caching off
+    """
+    def __init__(self, usecache=True, expiry_time=False, max_size=None):
```

Comment:
Is this max_size in terms of elements or bytes?

Evicting based on the number of elements doesn't seem especially useful to me, at least no compared to the size of the values. But perhaps s3fs or gcsfs already does that?

Reply:
It is number of elements as it stands, because it's far easier to implement using LRU. 
There is no such expiry anywhere yet; gcsfs has time-based expiry.